### PR TITLE
feat!: option to add row count on tables (#284)

### LINKED ISSUE #284: Option to add row count on tables
See [Figma](https://www.figma.com/design/g60pHZLz78wfl0PvbksMb5/HCA-Tracker?node-id=303-12855&t=LpMLTc9PY5JvwX0Y-0) and [background](https://github.com/clevercanary/cc-design/issues/143).

image

### PR DESCRIPTION
**Pull request by frano-m**
Closes #284.

image

### CODE REVIEW

File: src/components/Table/components/TableCell/components/RowPositionCell/rowPositionCell.tsx [hunk 1]
```diff
@@ -0,0 +1,16 @@
...
+import { BaseComponentProps } from "../../../../../types";
+import { TYPOGRAPHY_PROPS } from "./constants";
+
+export const RowPositionCell = <TData extends RowData, TValue>({
+  className,
+  ...cellContext
+}: BaseComponentProps & CellContext<TData, TValue>): JSX.Element => {
+  return (
+    <Typography {...TYPOGRAPHY_PROPS} className={className} component="div">
+      {cellContext.row.getRowPosition()}
```
Comment (frano-m) [inline src/components/Table/components/TableCell/components/RowPositionCell/rowPositionCell.tsx:13]: `RowPosition` is now a custom feature, and we now have `getRowPostiion` on the `cellContext.row`.

File: src/components/Detail/components/Table/table.tsx [hunk 2]
```diff
@@ -56,8 +58,11 @@ export const Table = <T extends RowData>({
   const { stickyHeader = false } = table || {};
   const { sx: tableContainerSx } = tableContainer || {};
   const tableInstance = useReactTable({
-    _features: [ROW_PREVIEW],
-    columns: generateColumnDefinitions(columns),
+    _features: [ROW_POSITION, ROW_PREVIEW],
```
Comment (frano-m) [inline src/components/Detail/components/Table/table.tsx:61]: New custom feature `ROW_POSITION`.

File: src/components/Table/features/RowPosition/constants.ts [hunk 3]
```diff
@@ -0,0 +1,34 @@
...
+  RowModel,
+  Table,
+  TableFeature,
+  TableState,
+} from "@tanstack/react-table";
+import { InitialTableState } from "@tanstack/table-core/src/types";
+import { getRowModel, getRowPosition, initInitialState } from "./utils";
+
+export const ROW_POSITION: TableFeature = {
+  createCell: <T extends RowData, TValue>(
```
Comment (frano-m) [inline src/components/Table/features/RowPosition/constants.ts:15]: We extend `cellContext` with the function on `row`: `getRowPosition`.

File: src/components/Table/features/RowPosition/constants.ts [hunk 4]
```diff
@@ -0,0 +1,34 @@
...
+    };
+  },
+  createTable: <T extends RowData>(table: Table<T>): void => {
+    const originalGetRowModel = table.getRowModel.bind(table);
+    table.getRowModel = (): RowModel<T> => {
+      return getRowModel(table, originalGetRowModel);
+    };
+  },
+  getInitialState: (initialState?: InitialTableState): Partial<TableState> => {
+    return initInitialState(initialState);
```
Comment (frano-m) [inline src/components/Table/features/RowPosition/constants.ts:32]: The default initial state will be to hide the row position column.

File: src/components/Table/features/RowPosition/utils.ts [hunk 5]
```diff
@@ -0,0 +1,80 @@
...
+ * @param getRowModel - Table getRowModel function.
+ * @returns row model.
+ */
+export function getRowModel<T extends RowData>(
+  table: Table<T>,
+  getRowModel: Table<T>[`getRowModel`]
+): RowModel<T> {
+  const rowModel = getRowModel();
+  rowModel.rows.forEach(({ id }, i) => {
+    rowModel.rowsById[id].getRowPosition = (): number =>
```
Comment (frano-m) [inline src/components/Table/features/RowPosition/utils.ts:23]: Adding the function `getRowPosition` onto `rowModel.rowsById[id]` where `rowsById` is part of TanStack's rowModel API, which maps each `row` to its `id`.

File: src/components/Table/features/RowPosition/utils.ts [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+  initialState?: InitialTableState
+): Partial<TableState> {
+  return {
+    ...initialState,
+    columnVisibility: {
+      [ACCESSOR_KEYS.ROW_POSITION]: false,
+      ...initialState?.columnVisibility,
+    },
+    pagination: {
+      ...DEFAULT_PAGINATION,
```
Comment (frano-m) [inline src/components/Table/features/RowPosition/utils.ts:76]: Custom feature initial state expects pagination to be defined `pageIndex` and `pageSize`. These values are initial state only, and will always be updated by table state as per usual.

**Review by NoopDog — APPROVED** [review]
Awesome! Thanks @frano-m !

**Pull request merged**